### PR TITLE
Add performance unit test for timeline-chart

### PR DESCRIPTION
**Pull request by hoangphamEclipse**
This commit adds unit tests that measure the performance of the bottlenecks of the timeline chart while zooming in. With these unit tests, developers now can measure the effects of their patches on the performance of the timeline chart.

### CODE REVIEW

File: timeline-chart/src/layer/unitTest/data/longRemoval/settings.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+            timeGraphRowIds: timeGraphRowIds,
+            timeGraphStates: timeGraphStates
+        },
+        {
+            timeGraphEntries: timeGraphEntries,
+            timeGraphRowIds: timeGraphRowIds,
+            timeGraphStates: timeGraphStates2
+        }
+    ]
+}
```
Comment (bhufmann) [inline timeline-chart/src/layer/unitTest/data/longRemoval/settings.ts:31]: No newline at end of file

File: timeline-chart/src/layer/unitTest/data/longRemoval/test-ids.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        37,
+        38,
+        39,
+        40,
+        41,
+        42,
+        43,
+        44
+    ] 
+}
```
Comment (bhufmann) [inline timeline-chart/src/layer/unitTest/data/longRemoval/test-ids.ts:49]: No newline at end of file

File: timeline-chart/src/layer/unitTest/data/noLabels/settings.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+        end: BigInt("719079351")
+    },
+    data: [
+        {
+            timeGraphEntries: timeGraphEntries,
+            timeGraphRowIds: timeGraphRowIds,
+            timeGraphStates: timeGraphStates
+        }
+    ]
+}
```
Comment (bhufmann) [inline timeline-chart/src/layer/unitTest/data/noLabels/settings.ts:25]: No newline at end of file

File: timeline-chart/src/layer/unitTest/data/noLabels/test-ids.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        4736,
+        4737,
+        4738,
+        4739,
+        4740,
+        4741,
+        4742,
+        4743
+    ]  
+}
```
Comment (bhufmann) [inline timeline-chart/src/layer/unitTest/data/noLabels/test-ids.ts:49]: No newline at end of file

File: timeline-chart/src/layer/unitTest/data/withLabels/settings.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+        end: BigInt("3027146240")
+    },
+    data: [
+        {
+            timeGraphEntries: timeGraphEntries,
+            timeGraphRowIds: timeGraphRowIds,
+            timeGraphStates: timeGraphStates
+        }
+    ]
+}
```
Comment (bhufmann) [inline timeline-chart/src/layer/unitTest/data/withLabels/settings.ts:25]: No newline at end of file

File: timeline-chart/src/layer/unitTest/data/withLabels/test-ids.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        37,
+        38,
+        39,
+        40,
+        41,
+        42,
+        43,
+        44
+    ] 
+}
```
Comment (bhufmann) [inline timeline-chart/src/layer/unitTest/data/withLabels/test-ids.ts:49]: No newline at end of file

File: timeline-chart/src/layer/unitTest/utils.ts [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+        timeGraphStates: any;
+    }
+
+    export interface TimeGraphPerformanceTestDataStub {
+        toNextStub(): void;
+        getEntries(): TimeGraphEntry[];
+        getIds(): number[];
+        getRows(): TestData.TimeGraphRow[];
+    }
+}
```
Comment (bhufmann) [inline timeline-chart/src/layer/unitTest/utils.ts:166]: No newline at end of file

File: timeline-chart/src/layer/__tests__/time-graph-chart-long-removal-test.ts [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    it ('addOrUpdateRow() test with long removal time', () => {
+        const data = timeGraph.getData();
+        const timeGraphChart = timeGraph.getTimeGraphChart();
+
+        const start = performance.now();
+        // @ts-ignore
+        timeGraphChart.addOrUpdateRows(data);
+        const end = performance.now();
+        const time = end - start;
+        console.log("addOrUpdateRow() - with long removal time", time);
```
Comment (bhufmann) [inline timeline-chart/src/layer/__tests__/time-graph-chart-long-removal-test.ts:27]: you should add the unit to the print-out.

File: timeline-chart/src/layer/__tests__/time-graph-chart-with-labels.test.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+        const data = timeGraph.getData();
+
+        const start = performance.now();
+        // We don't need to fetch any new data at all, the function should wipe out all existing rows
+        // and rerender them.
+        // @ts-ignore
+        timeGraphChart.addOrUpdateRows(data);
+        const end = performance.now();
+        const time = end - start;
+        console.log("addOrUpdateRow() - with labels running time", time);
```
Comment (bhufmann) [inline timeline-chart/src/layer/__tests__/time-graph-chart-with-labels.test.ts:46]: you should add the unit to the print-out.

File: timeline-chart/src/layer/__tests__/time-graph-chart-with-labels.test.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+        // Set the zoom range so that we don't have to call adjustZoom()
+        timeGraph.setViewRange(WithLabelTestData.zoomRange.start, WithLabelTestData.zoomRange.end); 
+
+        const start = performance.now();
+        // @ts-ignore
+        timeGraphChart.updateScaleAndPosition();
+        const end = performance.now();
+        const time = end - start;
+        console.log("updateScaleAndPosition() - with labels running time", time);
```
Comment (bhufmann) [inline timeline-chart/src/layer/__tests__/time-graph-chart-with-labels.test.ts:31]: you should add the unit to the print-out.

File: timeline-chart/src/layer/__tests__/time-graph-chart-without-labels.test.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+
+        // Set the zoom range so that we don't have to call adjustZoom()
+        timeGraph.setViewRange(NoLabelTestData.zoomRange.start, NoLabelTestData.zoomRange.end); 
+
+        const start = performance.now();
+        // @ts-ignore
+        timeGraphChart.updateScaleAndPosition();
+        const end = performance.now();
+        const time = end - start;
+        console.log("updateScaleAndPosition() - without labels running time", time);
```
Comment (bhufmann) [inline timeline-chart/src/layer/__tests__/time-graph-chart-without-labels.test.ts:26]: you should add the unit to the print-out.

File: timeline-chart/src/layer/__tests__/time-graph-chart-without-labels.test.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+        const data = timeGraph.getData();
+
+        const start = performance.now();
+        // We don't need to fetch any new data at all, the function should wipe out all existing rows
+        // and rerender them.
+        // @ts-ignore
+        timeGraphChart.addOrUpdateRows(data);
+        const end = performance.now();
+        const time = end - start;
+        console.log("addOrUpdateRow() - without labels running time", time);
```
Comment (bhufmann) [inline timeline-chart/src/layer/__tests__/time-graph-chart-without-labels.test.ts:41]: you should add the unit to the print-out.

File: timeline-chart/src/layer/unitTest/test-data-provider.ts [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+        // this.timeGraphIds = timeGraphIds;
+        // this.timeGraphEntries = timeGraphEntries;
+        // this.timeGraphStates = timeGraphStates;
+        this.data = data;
+        this.index = 0;
+        this.currentModel = this.data[this.index];
+    }
+
+    toNextStub(): void {
+        console.log("To next stub");
```
Comment (bhufmann) [inline timeline-chart/src/layer/unitTest/test-data-provider.ts:23]: is printout on purpose?

File: timeline-chart/src/layer/unitTest/test-data-provider.ts [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+        if (this.index + 1 < this.data.length) {
+            this.index++;
+            this.currentModel = this.data[this.index];
+            
+            console.log("New stub index", this.currentModel);
```
Comment (bhufmann) [inline timeline-chart/src/layer/unitTest/test-data-provider.ts:28]: is printout on purpose?

File: timeline-chart/src/layer/unitTest/test-data-provider.ts [hunk 15]
```diff
@@ -0,0 +1,211 @@
...
+                        },
+                        prevPossibleState,
+                        nextPossibleState,
+                        gapStyle
+                    });
+                }
+            }
+        });
+
+        console.log("Total number of states", totalStates);
```
Comment (bhufmann) [inline timeline-chart/src/layer/unitTest/test-data-provider.ts:163]: space after "states"

File: timeline-chart/src/layer/unitTest/test-data-provider.ts (hunk 13 above)
Comment (hoangphamEclipse) [reply timeline-chart/src/layer/unitTest/test-data-provider.ts]: No it was a mistake. I will remove it.

File: timeline-chart/src/layer/unitTest/test-data-provider.ts (hunk 15 above)
Comment (hoangphamEclipse) [reply timeline-chart/src/layer/unitTest/test-data-provider.ts]: I removed the comment, it was a mistake.

**Review by bhufmann — APPROVED** [review]
Code looks good. It will be useful to verify performance improvements. 

As I understood, the plan is to merge it late together with performance improvements, is that correct?

**Review by bhufmann — CHANGES_REQUESTED** [review]
Change looks good.  In the commit message you stated that developers can measure the effect of their changes. Could you add a sentence in the root README.md under chapter `Tests` (new sub chapter `Performance tests` on how the developer is supposed to do it? The test itself only prints-out the results and doesn't compare it an actual value, because the threshold would not be fix depending on what environment it runs.

**Comment by hoangphamEclipse** [comment]
Sounds good. I will add that.

File: README.md [hunk 16]
```diff
@@ -44,6 +44,25 @@ cd timeline-chart
...
+### Performance tests
+
+Developers can use unit tests to measure the performance improvement of their changes. The following section is a quick guide on how to set up unit tests.
+
+#### Set up
+
+The `TimeGraphPerformanceTest` is a class that conveniently creates a timeline chart for performance tests. It also provides some methods that return important
+test data, such as the total length of a trace for assertions. The code is taken from the timeline chart example.
+
+To construct a test, trace data and its corresponding view range is required. Developers can use POSTMAN or curl to obtain this data by sending a request to the server.
```
Comment (bhufmann) [inline README.md:56]: Please remove POSTMAN.

**Review by bhufmann — CHANGES_REQUESTED** [review]

**Review by bhufmann — APPROVED** [review]
LGTM. Thanks!

**Pull request merged**